File: src/Python/101-200/173.BinarySearchTreeIterator.py

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class BSTIterator:

    stack = []

    def __init__(self, root: TreeNode):
        if root is None:
            return
        self.stack.append(root)
        while root.left is not None:
            self.stack.append(root.left)
            root = root.left


    def next(self) -> int:
        """
        @return the next smallest number
        """
        node = self.stack.pop()
        if node.right is not None:
            self.stack.append(node.right)
            t = node.right
            while t.left is not None:
                self.stack.append(t.left)
                t = t.left
        return node.val


    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return len(self.stack) != 0


if __name__ == "__main__":
    obj = BSTIterator(root)
    param_1 = obj.next()
    param_2 = obj.hasNext()

File: src/Java/101-200/173.BinarySearchTreeIterator.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;


class BSTIterator {
    private LinkedList <TreeNode> stack = new LinkedList<TreeNode>();
    public BSTIterator(TreeNode root) {
        if (root == null){
            return;
        }
        stack.push(root);
        while (root.left != null){
            stack.push(root.left);
            root = root.left;
        }
    }

    /**
     * @return the next smallest number
     */
    public int next() {
        TreeNode node = this.stack.pop();
        if (node.right != null){
            this.stack.push(node.right);
            TreeNode t = node.right;
            while (t.left != null){
                stack.push(t.left);
                t = t.left;
            }
        }
        return node.val;
    }

    /**
     * @return whether we have a next smallest number
     */
    public boolean hasNext() {
        return !this.stack.isEmpty();
    }
}

